sklearn/cluster/tests/test_k_means.py

```diff
@@ -68,8 +70,70 @@ def test_kmeans_results(representation, algo, dtype):
     assert kmeans.n_iter_ == expected_n_iter
 
 
+@pytest.mark.parametrize("array_constr",
+                         [np.array, sp.csr_matrix],
+                         ids=['dense', 'sparse'])
+@pytest.mark.parametrize("algo", ['full', 'elkan'])
+def test_relocated_clusters(array_constr, algo):
+    # check that empty clusters are relocated as expected
+    X = array_constr([[0, 0], [0.5, 0], [0.5, 1], [1, 1]])
+
+    # second center too far from others points will be empty at first iter
+    init_centers = np.array([[0.5, 0.5], [3, 3]])
+
+    expected_labels = [0, 0, 1, 1]
+    expected_inertia = 0.25
+    expected_centers = [[0.25, 0], [0.75, 1]]
+    expected_n_iter = 3
+
+    kmeans = KMeans(n_clusters=2, n_init=1, init=init_centers, algorithm=algo)
+    kmeans.fit(X)
+
+    assert_array_equal(kmeans.labels_, expected_labels)
+    assert_almost_equal(kmeans.inertia_, expected_inertia)
+    assert_array_almost_equal(kmeans.cluster_centers_, expected_centers)
+    assert kmeans.n_iter_ == expected_n_iter
+
+
+@pytest.mark.parametrize("representation", ["dense", "sparse"])
+def test_relocate_empty_clusters(representation):
+    # test for the _relocate_empty_clusters_(dense/sparse) helpers
+
+    # Synthetic dataset with 3 obvious clusters of different sizes
+    X = np.array(
+        [-10., -9.5, -9, -8.5, -8, -1, 1, 9, 9.5, 10]).reshape(-1, 1)
+    if representation == "sparse":
+        X = sp.csr_matrix(X)
+    sample_weight = np.full(shape=10, fill_value=1.)
+
+    # centers all initialized to the first point of X
+    centers_old = np.array([-10., -10, -10]).reshape(-1, 1)
+
+    # With this initialization, all points will be assigned to the first center
+    # At this point a center in centers_new is the weighted sum of the points
+    # it contains if it's not empty, otherwise it is the same as before.
+    centers_new = np.array([-16.5, -10, -10]).reshape(-1, 1)
+    weight_in_clusters = np.array([10., 0, 0])
+    labels = np.zeros(10, dtype=np.int32)
+
+    if representation == "dense":
+        _relocate_empty_clusters_dense(X, sample_weight, centers_old,
+                                       centers_new, weight_in_clusters, labels)
+    else:
+        _relocate_empty_clusters_sparse(X.data, X.indices, X.indptr,
+                                        sample_weight, centers_old,
+                                        centers_new, weight_in_clusters,
+                                        labels)
+
+    # The relocation scheme will take the 2 points farthest from the center and
+    # assign them to the 2 empty clusters, i.e. points at 10 and at 9.9. The
+    # first center will be updated to contain the other 8 points.
+    assert_array_equal(weight_in_clusters, [8, 1, 1])
+    assert_allclose(centers_new, [[-36], [10], [9.5]])
+
+
 @pytest.mark.parametrize('distribution', ['normal', 'blobs'])
-@pytest.mark.parametrize('tol', [0, 1e-2, 1e-4, 1e-8])
+@pytest.mark.parametrize('tol', [1e-2, 1e-4, 1e-8])
```

Comment:
they can't be the same anymore when tol is 0? Looks like this is reverting a change frrom https://github.com/scikit-learn/scikit-learn/pull/16075

Reply:
Even with #16075 they are not guaranteed to be the same due to rounding errors. When you want to check that `||centers_new - centers_old|| <= tol=0`, you might not converge at the right iteration because a single rounding error in the computation of the norm results in a strictly positive number.

Since a rounding error might happen at a different iteration for elkan and lloyd, we can't really expect their n_iter to be equal. The test actually was
```py
# The number of iterations and inertia should be close but not
# necessarily exactly the same because of rounding errors.
assert km_elkan.n_iter_ == pytest.approx(km_full.n_iter_, rel=0.01)
```
Which is not very satisfying and the .01 relative error only works for this dataset. In practice the difference can be arbitrarily large. So instead I removed the 0 tol for this test and added a test that simply checks that convergence eventually happens with a zero tol, which is what #16075 fixed.

In addition, I mentioned in the docstring that tol=0 should be avoided and one should provide a very small value instead.

Reply:
OK Thanks. One slight confusion I have is that it seems that the test with `tol=0` was passing before (so it would probably still pass) since the check was

`assert km_elkan.n_iter_ == pytest.approx(km_full.n_iter_, rel=0.01)`

which is equivalent to a strict equality check since `n_iter_` is an int (you properly updated the check).

Anyway, I'm fine removing it considering the addition of the other test.

Reply:
`assert km_elkan.n_iter_ == pytest.approx(km_full.n_iter_, rel=0.01)` does check that the 2 ints are close in the sense that `|i1 - i2| < rel * i2`. The fact that they are integers does not make it a bare assert. Actually it used to pass because the difference of n_iter was 2 iirc. With this PR it jumped to something like 8 but it's only due to the number of iteration it takes after convergence to get exactly 0 when computing the displacement of the centers.